contracts/README.md

```diff
@@ -3,15 +3,396 @@
-- [PermissionedRegistry](src/common/PermissionedRegistry.sol) &mdash; an implementation of the v2 registry.
-- [UniversalResolver](src/universalResolver/UniversalResolver.sol) &mdash; onchain ENSv2 resolution.
-- [ETHFallbackResolver](src/L1/ETHFallbackResolver.sol) &mdash; crosschain resolver that combines mainnet v2 (ejected), mainnet v1 (unmigrated), and Namechain v2.
+- [Overview](#overview)
+- [Architecture](#architecture)
+  - [Core Concepts](#core-concepts)
+  - [Token ID System](#token-id-system)
+  - [Access Control](#access-control)
+- [Contract Documentation](#contract-documentation)
+  - [Registry System](#registry-system)
+  - [L2 Components](#l2-components)
+  - [L1 Components](#l1-components)
+  - [Cross-Chain Bridge](#cross-chain-bridge)
+  - [Resolution](#resolution)
+- [Getting started](#getting-started)
+- [Running the Devnet](#running-the-devnet)
+- [Miscellaneous](#miscellaneous)
+
+## Overview
+
+ENSv2 transitions from a flat registry to a hierarchical system that enables:
+
+- **L2 Scalability**: .eth names registered and managed on Layer 2 with reduced gas costs
+- **Flexible Ownership**: Custom registry implementations for different ownership models
+- **Cross-Chain Support**: Seamless name resolution across L1 and L2
+- **Backward Compatibility**: Unmigrated ENSv1 names continue to function
+- **Enhanced Security**: Names can be "ejected" to L1 for maximum security guarantees
+
+### Key Features
+
+1. **Hierarchical Registries**: Each name has its own registry contract managing its subdomains
+2. **Canonical ID System**: Stable storage identifiers enabling token regeneration
+3. **Bitmap-Based Access Control**: Gas-efficient permission management supporting up to 32 roles
+4. **Universal Resolver**: Single entry point for all name resolution
+5. **Migration Framework**: Transition path from ENSv1 to ENSv2
+
+## Architecture
+
+### Core Concepts
+
+**Registries**
+- Each registry is responsible for one name and its direct subdomains
+- Registries implement ERC1155, treating subdomains as NFTs
+- Must implement the `IRegistry` interface for standard resolution
+- All registries store data in the singleton `RegistryDatastore`
+
+**Root Registry** → **TLD Registries** (.eth, .box, etc.) → **Domain Registries** (example.eth) → **Subdomain Registries** (sub.example.eth)
+
+**Resolution Process**
+1. Start at root registry
+2. Recursively traverse to find the deepest registry with a resolver set
+3. Query that resolver for the requested record
+4. Supports wildcard resolution (parent resolver handles subdomains)
+
+### Token ID System
+
+ENSv2 uses a dual-identifier system to enable token regeneration while maintaining stable storage:
+
+#### How Token IDs Work
+
+```solidity
+// Label to Token ID conversion
```

Comment:
Remove this block, users don't need to know the format of token ids - they should use `LibLabel` instead.